docs/components/connectors/custom-built-connectors/best-practices.md

```diff
@@ -0,0 +1,182 @@
+
+**General Rule**: Begin with a lowercase letter and capitalize subsequent words (lower camel case). Ensure variables do not start with an underscore ("\_").
+
+For example:
+
+- ❌ \_MyTestVariable
+- ✅ myTestVariable
+
+**Exception for Method-Specific Properties**: When variables indicate method-specific properties, underscores are acceptable to separate the method indication from the actual property name. This enhances clarity, especially when the same properties are pre-populated for different methods.
+
+For example:
+
+- ✅ chatCompletion_apiVersion
+- ✅ completion_apiVersion
+
+#### Versioning
+
+Always include a version number property, starting from 1. Increment this number with each subsequent version.
+
+```json
+{
+  "$schema": "https://unpkg.com/@camunda/zeebe-element-templates-json-schema/resources/schema.json",
+  "name": "My Connector",
+  "id": "my.connector",
+  "version": 1
+}
+```
+
+#### JSON structure
+
+The JSON structure should mirror the visual presentation to maintain alignment with UI order. This ensures groups and properties are organized accordingly, even if some elements are hidden. This facilitates easier third-party maintenance.
+
+Note that if you want to use a FEEL expression that references other variables, the properties that define these variables in the JSON structure must be declared before the property where you intend to use the FEEL expression.
+
+#### Secret Handling
+
+- Follow the rules of the system you're adding to.
+- Don't save secret info in the JSON.
+- Use the Connectors' [secret handling](/components/connectors/use-connectors/index.md#using-secrets) instead.
+
+#### Field Types
+
+To ensure a good user experience, make sure the template is intuitive and designed to minimize errors.
+
+##### Hidden vs. visible fields
+
+A general rule of thumb is that if certain values are static and do not require user input, the template creator should prefill them. Since these values offer no additional benefit to the user by being visible, they can be hidden.
+
+Hidden fields - Example:
+
+- HTTP Method
+- Static URL Endpoint
+- Static header
+
+##### Required vs. optional
+
+- Required: All properties essential for the Element template to function. Without these properties, the template cannot operate.
+- Optional: All properties that are not mandatory but provide additional functionality.
+
+There are two mechanisms to define the connector property behavior depending on whether the field is required or optional:
+
+- [“Not empty” constraint](/components/modeler/element-templates/template-properties.md#validating-user-input-constraints): if activated, will display an error message if the user doesn’t fill in the field.
+- [Optional bindings](/components/modeler/element-templates/template-properties.md#preventing-persisting-empty-values-optional): an optional binding will not be persisted in the BPMN XML if the corresponding property contains an empty string.
```

Comment:
💭 When refactoring that part of the docs, I forgot to clear this up and I don't know it myself. Why would one ever actually need this?

Reply:
🤷

Reply:
how to make this comment actionable? Should we investigate? Should we remove the optional binding part? Should we just leave it because it's actually supported anyway?

Reply:
It's a thinking-out-loud comment. What I mean to say by that is you can decide to make it actionable and investigate, if you want to, but you don't have to if you don't think it's important enough (I don't think it's important or I'd have marked it as 🔧 :D)

Reply:
thanks for clarifying, I'll wait for Christina as a 3rd point of view :)

Reply:
Thanks for looping me in! My take:

- The optional binding is still a valid feature and supported in the product.  
- In practice, it’s mainly useful if you want a property to be truly optional and avoid persisting empty strings in the BPMN XML. For most cases, the “Not empty” constraint covers the common validation needs.  
- Given that, I’d recommend keeping the note about optional bindings in the docs for completeness, but we could add a brief clarification about when it’s actually useful.  

This way the docs stay accurate without overcomplicating things.